### PR TITLE
docs - add pl/container resource mgmt info

### PR DESCRIPTION
**Pull request by lisakowen**
- created a new subtopic "About PL/Container Resource Management" - content:
    - default resource management in pl/container
    - using resource groups to manage pl/container runtimes
    - includes a procedure to configure pl/container to use resource groups 
- this PR assumes doc updates to the "Using Resource Groups" page that are proposed in https://github.com/greenplum-db/gpdb/pull/4888 and https://github.com/greenplum-db/gpdb/pull/4921.
- a few misc edits and fixes

question - the python and r images that i downloaded from pivnet use the tag "devel".  is this correct?  the docs say the tag is supposed to be the version number.

doc review site link:
- http://docs-gpdb-review-staging.cfapps.io/review/ref_guide/extensions/pl_container.html#topic_resmgmt

### CODE REVIEW

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 1]
```diff
@@ -67,6 +68,149 @@
           >https://github.com/greenplum-db/plcontainer</xref>.</p>
     </body>
   </topic>
+  <topic id="topic_resmgmt">
+    <title>About PL/Container Resource Management</title>
+    <body>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I would remove the first paragraph, and reword the last sentence of the paragraph and move it near the end of the second paragraph.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 2]
```diff
@@ -67,6 +68,149 @@
...
+      <p>Greenplum Database runs PL/Container user-defined functions in Docker
+        containers. The Docker containers and the Greenplum Database server share
+        CPU and memory resources on the same host. In the default case, Greenplum
+        Database is unaware of the resources consumed by PL/Container instances.</p>
+      <p>PL/Container consumes resources at two levels - the container level and
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: In second paragraph, Is the first sentence correct? I thought "runtime" was just set of configuration settings that is used by PL/Container.

My suggestions. 
I would remove the first sentence.
I would start the second sentence with
"You can manage memory and CPU usage of PL/Container Docker containers with the memory_mb and cpu_share settings ..."

I would add the information about PL/Container and Greenplum Database resource management to the next tolast sentence of this paragraph. Something like.
"The PL/Container management of memory and CPU used by Docker containers is independent of Greenplum Database resource management."

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 3]
```diff
@@ -67,6 +68,149 @@
...
+        container are constrained only by the <codeph>memory_mb</codeph> and
+        <codeph>cpu_share</codeph> settings that you configure for the PL/Container
+        runtime. <codeph>memory_mb</codeph> governs the memory resources available
+        to each container instance of the runtime. The <codeph>cpu_share</codeph>
+        setting identifies the relative weighting of CPU usage compared to other
+        containers. Refer to <xref href="#topic_ojn_r2s_dw" format="dita"/> 
+        for PL/Container configuration information.</p>
+      <p>You cannot by default restrict the number of executing PL/Container
+        container instances, nor can you restrict the amount of memory or CPU
+        resources that they collectively consume.</p>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: maybe remove this paragraph or maybe keep the number of containers information.
 "you cannot directly restrict the number of concurrent Docker containers PL/Container runs."

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 4]
```diff
@@ -67,6 +68,149 @@
...
+
+    </body>
+  <topic id="topic_resgroup">
+    <title>Using Resource Groups to Manage PL/Container Resources</title>
+    <body>
+      <p>In PL/Container version 1.2.0 and later, you can use Greenplum Database
+        resource groups to manage and limit the CPU and memory resources of 
+        PL/Container runtimes. For more information about enabling, configuring,
+        and using Greenplum Database resource groups, refer to <xref href="../../admin_guide/workload_mgmt_resgroups.xml" format="dita" scope="peer">Using Resource Groups</xref>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I would add the pl/container/GPDB mem. management info as the second sentence of the first paragraph.
Something like this.
"When using resource groups, memory and CPU management is better integrated with Greenplum Database."

I have some questions
Q: Is the term "PL/Container runtimes" correct? I thought it is Docker containers. I thought runtimes were just a set of configuration settings.
Q: When using resource groups, can or should the PL/Container configuration settings for CPU and memory be removed?
Q: What happens when both the PL/Container CPU and mem. settings and resource_group_id setting is set?

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 5]
```diff
@@ -67,6 +68,149 @@
...
+        in the <cite>Greenplum Database Administrator Guide</cite>.</p>
+
+      <note>If you do not explicitly configure resource groups for a PL/Container
+        runtime, its container instances run unconstrained as described above.</note>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I don't think this note is necessary.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 6]
```diff
@@ -67,6 +68,149 @@
...
+        assignment using the object identifier (oid) of the resource group. You
+        can determine the oid for a given resource group name from the
+        <codeph>pg_resgroup</codeph> system catalog table. For example, the
+        following query displays the oid of a resource group named
+        <codeph>plpy_run1_rg</codeph>:<codeblock>SELECT rsgname, oid FROM pg_resgroup WHERE rsgname='plpy_run1_rg';
+    rsgname   |  oid  
+--------------+-------
+ plpy_run1_rg | 16391
+(1 row)</codeblock></p>
+
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Based on the catalog table description, the groupid is the same as the OID. This query is a bit less confusing because it doesn't use the "hidden" column oid.
SELECT groupid, groupname FROM gp_toolkit.gp_resgroup_config WHERE groupname='plpy_run1_rg'

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 7]
```diff
@@ -67,6 +68,149 @@
...
+        container instances are subject to the memory limit and the CPU limit that
+        you configured for the group. If you decrease the memory limit of a
+        PL/Container resource group, queries executing in running container
+        instances in the group may fail with an out of memory error. If you drop
+        a PL/Container resource group while there are running container instances,
+        Greenplum Database kills the running containers.</p>
+    </body>
+  </topic>
+  <topic id="topic_resgroupcfg">
+    <title>Procedure</title>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I think a more descriptive title would be better, something like 
"How to Configure PL/Container to Use Resource Groups"

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 8]
```diff
@@ -67,6 +68,149 @@
...
+          Greenplum Database resource groups:</p><ol>
+          <li>If you have not already configured and enabled resource groups in
+            your Greenplum Database cluster, configure cgroups and enable Greenplum
+            Database resource groups as described in
+            <xref href="../../admin_guide/workload_mgmt_resgroups.xml#topic71717999" format="dita" scope="peer">Using Resource Groups</xref>
+            in the <cite>Greenplum Database Administrator Guide</cite>. If you have
+            previously configured and enabled resource groups in your cluster,
+            ensure that the Greenplum Database resource group
+            <codeph>gpdb.conf</codeph> cgroups configuration file includes a
+            <codeph>memory { }</codeph> block as described in the previous link.</li>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I would put the sentence "If you have previously..." in a note for higher visibility.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 9]
```diff
@@ -67,6 +68,149 @@
...
+          <li>Analyze the resource usage of your Greenplum Database cluster.
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I think Greenplum Database "system" is the standard term. At least that's what I try to use.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 10]
```diff
@@ -67,6 +68,149 @@
...
+            Determine the amount of CPU and memory resources that you want to
+            allocate to PL/Container.<p><b>Result</b>: You identify the percentage
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: maybe rephrase "that you want to allocate to PL/Container"
to emphasize the resources are for Docker containers.
"...that you want to allocate to Docker containers that are run by PL/Container."

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 11]
```diff
@@ -67,6 +68,149 @@
...
+            of resource group memory and CPU resources for PL/Container.</p></li>
+          <li>Examine the PL/Container runtime definitions that you have configured
+            or plan to configure in your Greenplum Database deployment. Determine
+            if/how you want to distribute the global PL/Container CPU and memory
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I would change "if/how" to just "how"

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 12]
```diff
@@ -67,6 +68,149 @@
...
+            you created. For example:<codeblock>SELECT rsgname, oid FROM pg_resgroup WHERE rsgname='plpy_run1_rg';
+    rsgname   |  oid
+--------------+-------
+ plpy_run1_rg | 16391
+(1 row)
+SELECT rsgname, oid FROM pg_resgroup WHERE rsgname='plr_run1_rg';
+   rsgname   |  oid
+-------------+-------
+ plr_run1_rg | 16393
+(1 row)</codeblock></li>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: This might be a better query to display the group IDs, it displays both resource groups at once.
SELECT groupid, groupname FROM gp_toolkit.gp_resgroup_config WHERE groupname in ('plpy_run1_rg', 'plr_run1_rg')

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 13]
```diff
@@ -706,11 +850,16 @@ $$ LANGUAGE plcontainer;</codeblock></p>
...
-                      <li><codeph>cpu_share</codeph> - Set relative waiting of CPU usage CPU usage
-                        compared to other containers. The default value is 1024. The value is a
+                      <li><codeph>cpu_share</codeph> - Set the weighting of CPU usage
+                        for the runtime. The default value is 1024. The value is a
                         relative weighting of CPU usage compared to other containers.</li>
                       <li><codeph>memory_mb</codeph> - Set the memory allocated for the container.
+                        The default value is 1024.
                         The value is an integer that specifies the amount of memory in MB. </li>
+                      <li><codeph>resource_group_id</codeph> - Assign the specified resource group to the runtime.
+                        You must specify the object identifier of the resource group.
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I would change "object identifier" to "groupid".

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 14]
```diff
@@ -1035,6 +1185,15 @@ $$ LANGUAGE plcontainer;</codeblock></p>
                           means in case of memory overcommit, the container is terminated by the
                           system.</pd>
                       </plentry>
+                      <plentry>
+                        <pt>resource_group_id="<varname>rg_oid</varname>"</pt>
```
Comment (mkiyama) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I think "rg_oid" should be something like "res_grp_groupid"
Also in the description
--I think the description should use the res. group groupid instead of OID

--I would not say "run collectively unconstrained" I think something like this would be better
"PL/Container resource management is independent of Greenplum Database resource management"

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 2 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama  In second paragraph, the first sentence is a little confusing.  
"PL/Container consumes resources at two levels - the container level and runtime level"
better to replace by "PL/Container limits/manages the resource usage at two levels - the container level and runtime level"
Since we specify/config cgroup parent on each runtime to manage the memory/CPU usage for all the containers on that runtime, So I think to use runtime level is OK.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @zhangh43 
when you say "runtime" do you mean the set of Docker containers that are run by PL/Container and that use the settings specified in the runtime element of the pl/container configuation file?

thanks

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 4 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama 
Runtime is the configuration to describe the container type, and It is also one of the resource management unit.
There are two resource management unit/level: container level.(which is per container and use memory_mb setting to control) and runtime level(which apply on all the containers in that runtime and use resource_group_id setting to control)
You could see that there are three setting in runtime description
1 resource_group_id: control the total cpu and memory resource usage for all the containers in this runtime
2 memory_mb: control the memory limit for every container in this runtime
3 cpu_share:  control the cpu limit for every container in this runtime

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 6 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama good suggestion.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 13 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama  agree there are five places using "object identifier", please replace them together.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 2 above)
Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @zhangh43 - I am confused about the two levels of consumption, container and runtime.
Is container level the resources consumed by the docker container?
Is runtime level the resources consumed by the GPDB process that called the PL/Container UDF, or is it the the resources for the PL/Container process that runs the Docker container, or something else.

thanks

Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama 
1 Is container level the resources consumed by the docker container?
Answer: container level only limit resource usage for only on container. But there are many docker containers, using container level only cannot control the total resource usage of PLContainer component. So we introduce runtime level. It's used to control the resource usage of PLContainer component.
2 Is runtime level the resources consumed by the GPDB process that called the PL/Container UDF?
Answer: no, it also control resource usage of containers. Resource usage of GPDB process is always managed by normal resource group(not plcontainer resource group).

Here are summary.
There are two resource management unit/level: container level.(which is per container and use memory_mb setting to control) and runtime level(which apply on all the containers in that runtime and use resource_group_id setting to control)
You could see that there are three setting in runtime description
1 resource_group_id: control the total cpu and memory resource usage for all the containers in this runtime
2 memory_mb: control the memory limit for every container in this runtime
3 cpu_share: control the cpu limit for every container in this runtime

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama  yes.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 2 above)
Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Thank you for clarifying.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 4 above)
Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Thanks. It was not clear to me that the resource_group_id setting controls the resources for all the running PL/container containers that are associated with the runtime configuration.

Sorry for any confusion I caused.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 5 above)
Comment (mkiyama) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: for the phrase "its container instances run unconstrained as described above."

I would suggest something like.
"the number concurrent containers associated with the runtime configuration that PL/Container could run is only limited by system resources."

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 2 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Agree

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 5 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @mkiyama  yes, and I think this paragraph should also emphasize that lack of plcontainer resource group, memory usage of PLContainer cannot be controlled and PLContainer may consume resources which belong to GPDB core engine

**Comment by lisakowen** [comment]
thank you for reviewing, @zhangh43 .  i pushed a commit with edits, please let me know if i did not address your concerns.

**Comment by lisakowen** [comment]
@zhangh43 - can you please review for accuracy the note i added to the "Configuring Resource Groups for PL/Container" section regarding downgrade?

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 15]
```diff
@@ -67,6 +68,164 @@
           >https://github.com/greenplum-db/plcontainer</xref>.</p>
     </body>
   </topic>
+  <topic id="topic_resmgmt">
+    <title>About PL/Container Resource Management</title>
+    <body>
+      <p>Greenplum Database runs PL/Container user-defined functions in Docker
+        containers. The Docker containers and the Greenplum Database server share
+        CPU and memory resources on the same host. In the default case, Greenplum
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: host -> hosts

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 16]
```diff
@@ -67,6 +68,164 @@
...
+        Database is unaware of the resources consumed by running PL/Container instances.</p>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: I think it would be good to move the concluding sentence of this section to here, just to highlight the new 1.2 capabilities.  I think someone else had a similar comment.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 17]
```diff
@@ -67,6 +68,164 @@
...
+      <p>PL/Container manages resource usage at two levels - the container level and
+        the runtime level. You can control container-level CPU and memory resources
+        with the <codeph>memory_mb</codeph> and
+        <codeph>cpu_share</codeph> settings that you configure for the PL/Container
+        runtime. <codeph>memory_mb</codeph> governs the memory resources available
+        to each container instance of the runtime. The <codeph>cpu_share</codeph>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Seems like it would be clearer to remove "of the runtime" here.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 18]
```diff
@@ -67,6 +68,164 @@
...
+        setting identifies the relative weighting of CPU usage compared to other
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: "weighting of CPU usage" -> "weighting of a container's CPU usage"

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 19]
```diff
@@ -67,6 +68,164 @@
...
+      <p>In PL/Container version 1.2 and later, you can use Greenplum Database resource
+         groups to control overall CPU and memory resource usage for running PL/Container
+         instances, as described in the following section.</p>
+
+    </body>
+  <topic id="topic_resgroup">
+    <title>Using Resource Groups to Manage PL/Container Resources</title>
+    <body>
+      <p>In PL/Container version 1.2.0 and later, you can use Greenplum Database
+        resource groups to manage and limit the CPU and memory resources of
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: "of PL/Container runtimes" -> "of containers"

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 20]
```diff
@@ -67,6 +68,164 @@
...
+        <codeph>CONCURRENCY=0</codeph> in addition to the required CPU and memory
+        limits. For example, the following command creates a resource group named
+        <codeph>plpy_run1_rg</codeph> for a PL/Container runtime:
+        <codeblock>CREATE RESOURCE GROUP plpy_run1_rg WITH (MEMORY_AUDITOR=cgroup, CONCURRENCY=0,
+     CPU_RATE_LIMIT=10, MEMORY_LIMIT=10);</codeblock></p>
+
+      <p>PL/Container does not use the <codeph>MEMORY_SHARED_QUOTA</codeph> and
+        <codeph>MEMORY_SPILL_RATIO</codeph> resource group memory limits. Refer to
+        the <xref href="../../ref_guide/sql_commands/CREATE_RESOURCE_GROUP.xml" format="dita" scope="peer">CREATE RESOURCE GROUP</xref>
+        reference page for detailed information on this SQL command.</p>
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: on -> about

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 21]
```diff
@@ -67,6 +68,164 @@
...
+            Database resource groups as described in
+            <xref href="../../admin_guide/workload_mgmt_resgroups.xml#topic71717999" format="dita" scope="peer">Using Resource Groups</xref>
+            in the <cite>Greenplum Database Administrator Guide</cite>.<note>If you have
+            previously configured and enabled resource groups in your deployment,
+            ensure that the Greenplum Database resource group
+            <codeph>gpdb.conf</codeph> cgroups configuration file includes a
+            <codeph>memory { }</codeph> block as described in the previous link.</note></li>
+          <li>Analyze the resource usage of your Greenplum Database deployment.
+            Determine the amount of CPU and memory resources that you want to
+            allocate to PL/Container Docker containers.<p><b>Result</b>: You identify the percentages
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Not sure I like this **Result** construct.  I think it would be better to just edit the previous sentence to say "Determine the percentage of CPU and memory resources that you want to allocate to ..."

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 22]
```diff
@@ -67,6 +68,164 @@
...
+            of resource group memory and CPU resources for PL/Container.</p></li>
+          <li>Examine the PL/Container runtime definitions that you have configured
+            or plan to configure in your Greenplum Database deployment. Determine
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Here, better to just say "Determine:" and skip to the bullet list below.

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 23]
```diff
@@ -67,6 +68,164 @@
...
+            you created. For example:<codeblock>SELECT groupname, groupid FROM gp_toolkit.gp_resgroup_config 
+    WHERE groupname IN ('plpy_run1_rg', 'plr_run1_rg');
+
+  groupname   |  groupid
+--------------+----------
+ plpy_run1_rg |   16391
+ plr_run1_rg  |   16393
+(1 row)</codeblock></li>
+          <li>Assign each resource group that you created to the desired
+            PL/Container runtime(s). If you have not yet created the runtime
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: runtime(s) -> runtime definitions

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 24]
```diff
@@ -67,6 +68,164 @@
...
+            definition, use the <codeph>plcontainer runtime-add</codeph> command.
+            If the runtime already exists, use the
+            <codeph>plcontainer runtime-replace</codeph> or
+            <codeph>plcontainer runtime-edit</codeph> command to add the 
+            resource group assignment to the runtime configuration. For example:
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: configuration -> definition

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 25]
```diff
@@ -706,11 +865,20 @@ $$ LANGUAGE plcontainer;</codeblock></p>
                     the XML attribute of the <codeph><xref href="#topic_ojn_r2s_dw/plc_settings"
                         format="dita">settings</xref></codeph> element in the PL/Container
                     configuration file. These are valid parameters.<ul id="ul_dsz_j4w_gcb">
-                      <li><codeph>cpu_share</codeph> - Set relative waiting of CPU usage CPU usage
-                        compared to other containers. The default value is 1024. The value is a
+                      <li><codeph>cpu_share</codeph> - Set the CPU limit for each container
+                        of the runtime. The default value is 1024. The value is a
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: of the runtime -> in the runtime definition

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 26]
```diff
@@ -706,11 +865,20 @@ $$ LANGUAGE plcontainer;</codeblock></p>
...
                         relative weighting of CPU usage compared to other containers.</li>
-                      <li><codeph>memory_mb</codeph> - Set the memory allocated for the container.
+                      <li><codeph>memory_mb</codeph> - Set the memory limit for each container
+                        of the runtime. The default value is 1024.
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: same edit here

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 27]
```diff
@@ -706,11 +865,20 @@ $$ LANGUAGE plcontainer;</codeblock></p>
...
                         The value is an integer that specifies the amount of memory in MB. </li>
+                      <li><codeph>resource_group_id</codeph> - Assign the specified resource group to the runtime.
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: runtime -> runtime definition

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 28]
```diff
@@ -1019,7 +1188,8 @@ $$ LANGUAGE plcontainer;</codeblock></p>
                   <pd>These are the valid attributes.<parml>
                       <plentry>
                         <pt>cpu_share</pt>
-                        <pd>Optional. Specify the CPU usage for a PL/Container container. The value
+                        <pd>Optional. Specify the CPU usage for each PL/Container container
+                          in the runtime. The value
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: runtime -> runtime definition

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 29]
```diff
@@ -1028,13 +1198,27 @@ $$ LANGUAGE plcontainer;</codeblock></p>
...
                           limited by the host system <codeph>cgroups</codeph> configuration, which
                           means in case of memory overcommit, the container is terminated by the
                           system.</pd>
                       </plentry>
+                      <plentry>
+                        <pt>resource_group_id="<varname>rg_groupid</varname>"</pt>
+                        <pd>Optional. The value specifies the <codeph>groupid</codeph> of the resource
+                          group to assign to the PL/Container runtime.
+                          The resource group limits the total CPU and memory resource
+                          usage for all running containers that share this runtime
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: remove "running"

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml [hunk 30]
```diff
@@ -1028,13 +1198,27 @@ $$ LANGUAGE plcontainer;</codeblock></p>
...
+                          definition. You must specify the <codeph>groupid</codeph>
+                          of the resource group. If you do not assign a 
+                          resource group to a PL/Container runtime, its container instances
```
Comment (dyozie) [inline gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: runtime -> runtime definition

File: gpdb-doc/dita/ref_guide/extensions/pl_container.xml (hunk 19 above)
Comment (zhangh43) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: @dyozie how about "the CPU and memory resources of PL/Container runtimes" ->  "the total CPU and memory resources of containers in PL/Container runtime"

Comment (dyozie) [reply gpdb-doc/dita/ref_guide/extensions/pl_container.xml]: Sure, that works.

**Comment by dyozie** [comment]
Thanks everyone for reviewing this new content.  It looks like we're at a good point to merge this, but feel free to continue providing feedback if you notice any other changes that are required.
